fix(milvus): Enhanced Milvus VectorDB Instrumentation for Improved search Monitoring

divyapathak24 opened the pull request:
<!-- Thanks for submitting a PR! To make sure this gets merged quickly, make sure to check the following checkboxes. -->
Fixes #2808 

- [x] I have added tests that cover my changes.
- [x] If adding a new instrumentation or changing an existing one, I've added screenshots from some observability platform showing the change.
- [x] PR name follows conventional commits format: `feat(instrumentation): ...` or `fix(instrumentation): ...`.
- [ ] (If applicable) I have updated the documentation accordingly.


**New search attributes for single-vector searches**

<img width="1147" alt="Screenshot 2025-04-15 at 12 13 46 AM" src="https://github.com/user-attachments/assets/1a8041bc-3eb8-4fc3-8c52-4c9c140939f2" />

**New search attributes for multi-vector searches**
<img width="1185" alt="Screenshot 2025-04-15 at 12 11 23 AM" src="https://github.com/user-attachments/assets/be1bf18e-54db-4cbb-85ef-317dc0feb5df" />

**Search Events to log results:**
<img width="1211" alt="Screenshot 2025-04-15 at 12 11 39 AM" src="https://github.com/user-attachments/assets/5af4ed9e-f1e7-4b3c-b9e8-b50557c047bf" />


<!-- ELLIPSIS_HIDDEN -->

----

> [!IMPORTANT]
> Enhance Milvus VectorDB instrumentation with detailed search monitoring attributes and events, and add tests for single and multi-vector searches.
> 
>   - **Instrumentation Enhancements**:
>     - Add search duration and result status attributes in `_wrap()` in `wrapper.py`.
>     - Log search result events in `_add_search_result_events()` in `wrapper.py`.
>     - Add new search attributes like `MILVUS_SEARCH_RADIUS`, `MILVUS_SEARCH_METRIC_TYPE`, and `MILVUS_SEARCH_INDEX_TYPE` in `semconv_ai/__init__.py`.
>   - **Testing**:
>     - Add `test_search.py` to test single and multi-vector searches, verifying span attributes and events.
>     - Test different radius values to ensure varying result counts.
>   - **Misc**:
>     - Import `time` module in `wrapper.py` for measuring search duration.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=traceloop%2Fopenllmetry&utm_source=github&utm_medium=referral)<sup> for 3fe4f70605773bf0e3a583b587db3d0982f384db. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

nirga reviewed (commented):
Thanks @divyapathak24! Can you fix the broken CI?

divyapathak24 commented:
> Thanks @divyapathak24! Can you fix the broken CI?

Sure. I will fix them @nirga

divyapathak24 commented:
> > Thanks @divyapathak24! Can you fix the broken CI?
> 
> Sure. I will fix them @nirga

@nirga  could you trigger build again? Thanks!

galkleinman commented on packages/opentelemetry-instrumentation-milvus/opentelemetry/instrumentation/milvus/wrapper.py:
```diff
@@ -211,6 +228,49 @@ def _set_search_attributes(span, kwargs):
+        AISpanAttributes.MILVUS_SEARCH_RADIUS,
+        (
+            kwargs.get("search_params", {}).get("radius")
+            if kwargs.get("search_params") and "radius" in kwargs.get("search_params")
+            else None
```
isn't it redundant?

galkleinman commented on packages/opentelemetry-instrumentation-milvus/opentelemetry/instrumentation/milvus/wrapper.py:
```diff
@@ -211,6 +228,49 @@ def _set_search_attributes(span, kwargs):
+        AISpanAttributes.MILVUS_SEARCH_METRIC_TYPE,
+        (
+            kwargs.get("search_params", {}).get("metric_type")
+            if kwargs.get("search_params")
+            and "metric_type" in kwargs.get("search_params")
+            else None
```
same as above

galkleinman commented on packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv_ai/__init__.py:
```diff
@@ -194,7 +194,20 @@ class SpanAttributes:
+    MILVUS_SEARCH_INDEX_TYPE = "db.milvus.search.index_type"
     MILVUS_SEARCH_TIMEOUT = "db.milvus.search.timeout"
+    MILVUS_SEARCH_RESULT_COUNT = "db.milvus.search.result_count"
+    MILVUS_SEARCH_RESULT_STATUS = "db.milvus.search.status"
```
remove?

divyapathak24 replied:
Just to confirm, do you mean remove these lines entirely and prepare a separate PR on semconv_ai ?

galkleinman commented on packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv_ai/__init__.py:
```diff
@@ -194,7 +194,20 @@ class SpanAttributes:
+    MILVUS_SEARCH_RESULT_MIN_DISTANCE = "db.milvus.search.result_min_distance"
+    MILVUS_SEARCH_RESULT_MAX_DISTANCE = "db.milvus.search.result_max_distance"
+    MILVUS_SEARCH_RESULT_AVG_DISTANCE = "db.milvus.search.result_avg_distance"
+    MILVUS_SEARCH_DURATION_IN_MS = "db.milvus.search.duration_ms"
```
remove?

galkleinman reviewed (commented):
Hey @divyapathak24,

Thanks for submitting this one! please look at my comments - if you agree with them, please adjust (all occurrences of the things i mentioned, i didn't comment on all of them), otherwise let's discuss it here :)

divyapathak24 commented:
> Hey @divyapathak24,
> 
> Thanks for submitting this one! please look at my comments - if you agree with them, please adjust (all occurrences of the things i mentioned, i didn't comment on all of them), otherwise let's discuss it here :)

Hi @galkleinman 
The explicit attributes like `MILVUS_SEARCH_DURATION_IN_MS`, `MILVUS_SEARCH_RESULT_STATUS`, `RADIUS`, and `METRIC_TYPE` make Milvus search insights directly accessible, avoiding post-processing and assumptions about span internals.

galkleinman commented:
> > Hey @divyapathak24,
> > Thanks for submitting this one! please look at my comments - if you agree with them, please adjust (all occurrences of the things i mentioned, i didn't comment on all of them), otherwise let's discuss it here :)
> 
> Hi @galkleinman The explicit attributes like `MILVUS_SEARCH_DURATION_IN_MS`, `MILVUS_SEARCH_RESULT_STATUS`, `RADIUS`, and `METRIC_TYPE` make Milvus search insights directly accessible, avoiding post-processing and assumptions about span internals.

Hey, 

It isn't span internal, an no additional post-processing is needed, it's just the way OTEL works afaik... `RADIUS` and `METRIC_TYPE` are indeed unique to Milvus. But the duration of the action (in this case search) implies for all of the spans, and therefore implemented OOTB by start_time, end_time and duration... So it feels (to me at least) like a data duplication and deviation from the standard/protocol.

divyapathak24 commented:
> > > Hey @divyapathak24,
> > > Thanks for submitting this one! please look at my comments - if you agree with them, please adjust (all occurrences of the things i mentioned, i didn't comment on all of them), otherwise let's discuss it here :)
> > 
> > 
> > Hi @galkleinman The explicit attributes like `MILVUS_SEARCH_DURATION_IN_MS`, `MILVUS_SEARCH_RESULT_STATUS`, `RADIUS`, and `METRIC_TYPE` make Milvus search insights directly accessible, avoiding post-processing and assumptions about span internals.
> 
> Hey again @divyapathak24,
> 
> It isn't span internal, an no additional post-processing is needed, it's just the way OTEL works afaik... `RADIUS` and `METRIC_TYPE` are indeed unique to Milvus. But the duration of the action (in this case search) implies for all of the spans, and therefore implemented OOTB by start_time, end_time and duration... So it feels (to me at least) like a data duplication and deviation from the standard/protocol.
> 
> @nirga wdyt?

Hi @galkleinman @nirga,

Thanks for the feedback! I will remove the redundant attributes. Just wanted to confirm if everything else looks good to you both. Let me know if you have any other suggestions for other attributes.

divyapathak24 commented:
@nirga @galkleinman Can you please review the changes? I have removed redundant attributes and pushed changes and updated test cases accordingly. 
cc: @hk-bmi

nirga commented:
Thanks @divyapathak24 - can you sign the CLA? - https://github.com/traceloop/openllmetry/pull/2815#issuecomment-2802645987

divyapathak24 commented:
@nirga done. Can you re-trigger the build?

divyapathak24 commented:
> @nirga done. Can you re-trigger the build?

@nirga @galkleinman It seems my test cases are failing as it is not able to find the newly added attributes in semconv ai. How do I resolve this? By changing semconv-ai package version from 0.4.3 to 0.4.4 ?

nirga commented:
Yes @divyapathak24 :)

divyapathak24 commented:
@nirga @galkleinman I have opened a another PR https://github.com/traceloop/openllmetry/pull/2883 for semconv version update which has updated search attributes.
